sql: server panics during Walk of `UPDATE` with placeholders in CASE.

Running the bank example over pg_wire causes server to die:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x78 pc=0x4676b9e]

goroutine 92 [running]:
panic(0x501f780, 0xc8200100f0)
	/usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql/parser.MapArgs.SetInferredType(0xc820d246c0, 0x791ecd0, 0xc8207f4d80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/david/code/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:745 +0x19e
github.com/cockroachdb/cockroach/sql/parser.(*CaseExpr).TypeCheck(0xc82097ae40, 0xc820d246c0, 0x0, 0x0, 0x0, 0x0)
	/Users/david/code/go/src/github.com/cockroachdb/cockroach/sql/parser/type_check.go:129 +0x6aa
github.com/cockroachdb/cockroach/sql.(*planner).Update(0xc820140000, 0xc820d51130, 0x5f331e00, 0x0, 0x0, 0xc800000001)
	/Users/david/code/go/src/github.com/cockroachdb/cockroach/sql/update.go:161 +0x1767
```

--- a/sql/parser/type_check.go
+++ b/sql/parser/type_check.go
@@ -83,9 +83,10 @@ func (expr *BinaryExpr) TypeCheck(args MapArgs) (Datum, error) {
 
 // TypeCheck implements the Expr interface.
 func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
-	var dummyCond, dummyVal Datum
+	var dummyCond Datum
 
-	valArgType := DummyBool
+	// If expr.Expr is nil, the WHEN clauses contain boolean expressions.
+	condArgType := DummyBool
 	if expr.Expr != nil {
 		var err error
 		dummyCond, err = expr.Expr.TypeCheck(args)
@@ -93,16 +94,20 @@ func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
 			return nil, err
 		}
 		if _, ok := dummyCond.(DValArg); ok {
-			return nil, fmt.Errorf("unsupported case expression type: %s", dummyCond.Type())
+			return nil, fmt.Errorf("cannot infer type for parameter %s", dummyCond)
 		}
-		valArgType = dummyCond
+		condArgType = dummyCond
 	}
 
+	dummyVal := DNull
 	if expr.Else != nil {
 		var err error
 		dummyVal, err = expr.Else.TypeCheck(args)
 		if err != nil {
 			return nil, err
+		}
+		if _, ok := dummyVal.(DValArg); ok {
+			return nil, fmt.Errorf("cannot infer type for parameter %s", dummyVal)
 		}
 	}
 
@@ -111,7 +116,7 @@ func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
 		if err != nil {
 			return nil, err
 		}
-		if set, err := args.SetInferredType(nextDummyCond, valArgType); err != nil {
+		if set, err := args.SetInferredType(nextDummyCond, condArgType); err != nil {
 			return nil, err
 		} else if set != nil {
 			nextDummyCond = set
@@ -126,12 +131,15 @@ func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
 		if err != nil {
 			return nil, err
 		}
+		if _, ok := nextDummyVal.(DValArg); ok && dummyVal == DNull {
+			return nil, fmt.Errorf("cannot infer type for parameter %s", nextDummyVal)
+		}
 		if set, err := args.SetInferredType(nextDummyVal, dummyVal); err != nil {
 			return nil, err
 		} else if set != nil {
 			nextDummyVal = set
 		}
-		if dummyVal == nil || dummyVal == DNull {
+		if dummyVal == DNull {
 			dummyVal = nextDummyVal
 		} else if !(nextDummyVal == DNull || nextDummyVal.TypeEqual(dummyVal)) {
 			return nil, fmt.Errorf("incompatible value types %s, %s", dummyVal.Type(), nextDummyVal.Type())

--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -211,16 +211,18 @@ func TestPGPrepareFail(t *testing.T) {
 	defer db.Close()
 
 	testFailures := map[string]string{
-		"SELECT $1 = $1":                           "pq: unsupported comparison operator: <parameter> = <parameter>",
-		"SELECT $1 > 0 AND NOT $1":                 "pq: incompatible NOT argument type: int",
-		"SELECT $1":                                "pq: could not determine data type of parameter $1",
-		"SELECT $1 + $1":                           "pq: unsupported binary operator: <parameter> + <parameter>",
-		"SELECT now() + $1":                        "pq: unsupported binary operator: <timestamp> + <parameter>",
-		"SELECT CASE $1 WHEN 1 THEN 1 END":         "pq: unsupported case expression type: parameter",
-		"SELECT CASE 1 WHEN 2 THEN 1 ELSE $1 END":  "pq: incompatible value types parameter, int",
-		"SELECT CASE 1 WHEN 2 THEN $2 ELSE $1 END": "pq: could not determine data type of parameter $1",
-		"CREATE TABLE $1 (id INT)":                 "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
-		"UPDATE d.t SET s = i + $1":                "pq: value type int doesn't match type STRING of column \"s\"",
+		"SELECT $1 = $1":                                "pq: unsupported comparison operator: <parameter> = <parameter>",
+		"SELECT $1 > 0 AND NOT $1":                      "pq: incompatible NOT argument type: int",
+		"SELECT $1":                                     "pq: could not determine data type of parameter $1",
+		"SELECT $1 + $1":                                "pq: unsupported binary operator: <parameter> + <parameter>",
+		"SELECT now() + $1":                             "pq: unsupported binary operator: <timestamp> + <parameter>",
+		"SELECT CASE $1 WHEN 1 THEN 1 END":              "pq: cannot infer type for parameter $1",
+		"SELECT CASE WHEN TRUE THEN $1 END":             "pq: cannot infer type for parameter $1",
+		"SELECT CASE WHEN TRUE THEN $1 ELSE $2 END":     "pq: cannot infer type for parameter $2",
+		"SELECT CASE WHEN TRUE THEN 1 ELSE $1 END":      "pq: cannot infer type for parameter $1",
+		"UPDATE d.t SET d = CASE WHEN TRUE THEN $1 END": "pq: cannot infer type for parameter $1",
+		"CREATE TABLE $1 (id INT)":                      "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
+		"UPDATE d.t SET s = i + $1":                     "pq: value type int doesn't match type STRING of column \"s\"",
 	}
 
 	if _, err := db.Exec(`CREATE DATABASE d; CREATE TABLE d.t (i INT, s STRING, d INT)`); err != nil {
@@ -348,6 +350,10 @@ func TestPGPreparedQuery(t *testing.T) {
 		},
 		"SELECT (SELECT 1+$1)": {
 			base.Params(1).Results(2),
+		},
+		"SELECT CASE WHEN $1 THEN $2 ELSE 3 END": {
+			base.Params(true, 2).Results(2),
+			base.Params(false, 2).Results(3),
 		},
 		// TODO(mjibson): test date/time types
 

